File: PythonCode/Repo_CompIntroQP_Python/ChapterTD/Exercise_5_5_2_adiabatic.py
"""
 This script calculates the probability to remain in the the 
 initial state for a particle in a harmonic osciallater which
 is shaken. It does so by solving the Schrödinger equation formulated
 within the adiabatic basis. The time-stepping is performed
 using a propagator of Crank-Nicolson form.
 
 The harmonic oscillated potential is subject to a 
 time-dependent translation which shifts it in both directions
 before restoring it in its original position. The dynamics is
 resolved repeatedly with varying duration.

 The harmonic oscillator potential is subject to a time-dependent 
 translation which shifts it in both directions before restoring it 
 in its original position. In addition to time, it takes omega as an 
 input; the duraton of the time-dependent shift is related to omega by 
 T = 2*pi/omega.

 Inputs
 L         - size of domain 
 N         - number of grid points (should be 2^n)
 Nt        - the number of time steps for each run
 k         - the strength of the potential
 omegaMin  - minimal omega value
 omegaMax  - maximal omega value
 domega    - step size in omega
 InitialState - the index of the initial state; 
 InitialState = 0 is the ground state

 All inputs are hard coded initially
"""

# Libraries
import numpy as np
from matplotlib import pyplot as plt
from scipy import linalg


# Grid parameters
L = 10
N = 512              

# Strength of the harmonic oscillator
Kpot = 1

# Time parameters
Nstep = 500
omegaMin = 0.05
omegaMax = 0.5
domega = 0.025

# Index of initial state (ground state corresponds to zero)
InitialState = 0

# Shape of the potential
def Vpot(x):
    return 0.5*Kpot*x**2

# The displacement of the potential
def Ftrans(t, omega): 
    return 8/3**(3/2)*np.sin(omega*t/2)**2*np.sin(omega*t)

# The derivative of the displacement of the potential
def FtransDeriv(t, omega):
    return (t<2*np.pi/omega)*4/3**(3/2)*omega* \
    (2*np.sin(omega*t/2)**2*np.cos(omega*t) + np.sin(omega*t)**2)


# Set up grid
x = np.linspace(-L/2, L/2, N)
h = L/(N-1)

# Kinetic energy operator (FFT)
k_max = np.pi/h
dk = 2*k_max/N
k = np.append(np.linspace(0, k_max-dk, int(N/2)), 
              np.linspace(-k_max, -dk, int(N/2)))
# Transform identity matrix
Itrans = np.fft.fft(np.identity(N, dtype=complex), axis = 0)
# Multiply by (ik)^2 and k for T and P, respectively
Tmat = np.matmul(np.diag(-k**2), Itrans)
Pmat = np.matmul(np.diag(k), Itrans)
# Transform back to x-representation. 
Tmat = np.fft.ifft(Tmat, axis = 0)
Pmat = np.fft.ifft(Pmat, axis = 0)
# Correct pre-factor (unit mass and \hbar = 1)
Tmat = -1/2*Tmat    

# Add potential
# Full Hamiltonian
Ham = Tmat + np.diag(Vpot(x))

# Diagaonalize time-independent Hamiltonian (Hermitian matrix)
Evector, Bmat = np.linalg.eigh(Ham)
# Normalize eigenstates
Bmat = Bmat/np.sqrt(h)

# Digonal matrix with energies
Dmat = np.diag(Evector)
# Coupling matrix
aux = np.matmul(Pmat, Bmat)
CoupMat = h*np.matmul(np.conj(Bmat.T), aux)

# InitialState
Avec0 = np.zeros((N, 1))
Avec0[InitialState] = 1
Imat = np.identity(N) 

# Loop over the omega values
omegaVector = np.arange(omegaMin, omegaMax+domega/10, domega)
Pvector = 0*omegaVector          # Allocate    
index = 0
for omega in omegaVector:
  # Duration and time-step
  tTotal = 2*np.pi/omega
  dt = tTotal/Nstep
  t = 0
  Fderiv = 0
  Ham = Dmat
  Avec = Avec0
  # Do the dynamics
  while t < tTotal:
    # Update Hamiltonian and the derivative of the displament
    FderivOld = Fderiv;
    HamOld = Ham;
    Fderiv = FtransDeriv(t+dt, omega);
    Ham = Dmat - Fderiv*CoupMat;
    
    # The Crank-Nicolson scheme
    Avec = np.matmul((Imat - 1j*HamOld*dt/2), Avec)
    InvStep = np.linalg.inv(Imat + 1j*Ham*dt/2) 
    Avec = np.matmul(InvStep, Avec)
    
    # Update time
    t = t + dt
  # Probability for remaining in the initial state
  Pinit = np.abs(Avec[InitialState])**2
  Pinit = float(Pinit)
  # Write result to screen
  print(f'omega: {omega:.4f}, Pinit: {Pinit:.4f}')
  Pvector[index] = Pinit
  index = index + 1

# Plot Pinit as a function of omega
plt.figure(1)
plt.clf()
plt.plot(omegaVector, Pvector, '-', linewidth = 2, color = 'black')
plt.xlabel(r'$\omega$', fontsize = 12)
plt.ylabel(r'$P_{init}$', fontsize = 12)
plt.grid()
plt.show()

File: MatlabCode/Chapter5_Time-dependent/Exercise_5_5_2_adiabatic.m
% This script calculates the probability to remain in the the 
% initial state for a particle in a harmonic osciallater which
% is shaken. It does so by solving the Schrödinger equation formulated
% within the adiabatic basis. The time-stepping is performed
% using a propagator of Crank-Nicolson form.
% 
% The harmonic oscillated potential is subject to a 
% time-dependent translation which shifts it in both directions
% before restoring it in its original position. The dynamics is
% resolved repeatedly with varying duration.
%
% The harmonic oscillator potential is subject to a time-dependent 
% translation which shifts it in both directions before restoring it 
% in its original position. In addition to time, it takes omega as an 
% input; the duraton of the time-dependent shift is related to omega by 
% T = 2*pi/omega.
%
% Inputs
% L         - size of domain 
% N         - number of grid points (should be 2^n)
% Nt        - the number of time steps for each run
% k         - the strength of the potential
% omegaMin  - minimal omega value
% omegaMax  - maximal omega value
% domega    - step size in omega
% InitialState - the index of the initial state; 
% InitialState = 0 is the ground state
%
% All inputs are hard coded initially

% Grid parameters
L = 15;
N = 512;              

% Strength of the harmonic oscillator
k = 1;

% Time parameters
Nstep = 250;
omegaMin = 0.05;
omegaMax = 0.5;
domega = 0.05;

% Index of initial state (ground state corresponds to zero)
InitialState = 0;

% Shape of the potential
Vpot = @(x) 0.5*k*x.^2;

% The derivative of displacement of the potential
FtransDeriv = @(t, omega) (t<2*pi/omega)*4/3^(3/2)*omega*...
    (2*sin(omega*t/2).^2.*cos(omega*t) + sin(omega*t).^2);

% Set up the grid
x = linspace(-L/2,L/2,N);
x = x.';                            % Transpose
h = L/(N-1);

% Set up momentum and kinetic energy matrix
k = 2*pi/(N*h)*[0:(N/2-1), (-N/2:-1)];          % Vector with k-values
% Fourier transform the identity matrix:
Itrans = fft(eye(N));
% Multiply by (ik)^2 for T and just ik for p
Tmat = diag(-k.^2)*Itrans;
Pmat = diag(1i*k)*Itrans;
% Transform back to x-representation
Tmat = ifft(Tmat);
Tmat = -1/2*Tmat;            % Correct prefactor
Pmat = -1i*ifft(Pmat);

% Diagonalize initial Hamiltonian (in grid representation)
Hinit = Tmat + diag(Vpot(x));
[B E] = eig(Hinit);
E = diag(E);
% Sort and normalize
[E Ind] = sort(E);
B = B(:, Ind)/sqrt(h);

% Hamiltonian 
Dmat = diag(E);             % Diagonal energies (time-independent)
CoupMat = h*B'*Pmat*B;      % Coupling matrix

% Initiate and allocate
% Identity matrix
Imat = eye(N);
% Vectors with omega and Pinit values
omegaVector = omegaMin:domega:omegaMax;
Pvector = zeros(1, length(omegaVector));
index = 1;

% Loop over the omega values
for omega = omegaVector
  % Initial state
  Avec = zeros(N,1);
  Avec(InitialState + 1) = 1;
  
  % Time parameters
  tTotal = 2*pi/omega;
  dt = tTotal/Nstep;
  t = 0;
  Fderiv = 0;
  Ham = Dmat;
  
  % Do the dynamics
  while t < tTotal
    % Update Hamiltonian and the derivative of the displament
    FderivOld = Fderiv;
    HamOld = Ham;
    Fderiv = FtransDeriv(t+dt, omega);
    Ham = Dmat - Fderiv*CoupMat;
    % The Crank-Nicolson scheme
    Avec = (Imat - 1i*HamOld*dt/2)*Avec;
    Avec = inv(Imat + 1i*Ham*dt/2)*Avec;
    % Update time
    t = t + dt;
  end
  % Final probability
  Pinit = abs(Avec(InitialState+1)).^2;
  
  % Write propbability to screen
  disp(['omega: ', num2str(omega), ', Pinit: ', num2str(Pinit)])
  Pvector(index) = Pinit;
  index = index + 1;
% Plot Pinit as a function of omega
figure(1)
plot(omegaVector, Pvector, 'k-', 'linewidth', 2)
grid on
xlabel('\omega')
ylabel('P_{init}')
set(gca, 'fontsize', 15)
end

% Plot Pinit as a function of omega
figure(1)
plot(omegaVector, Pvector, 'k-', 'linewidth', 2)
grid on
xlabel('\omega')
ylabel('P_{init}')
set(gca, 'fontsize', 15)